common/src/main/java/com/microsoft/identity/common/internal/ui/embeddedwebview/AzureActiveDirectoryWebViewClient.java

```diff
@@ -0,0 +1,261 @@
+import com.microsoft.identity.common.internal.ui.embeddedwebview.challengehandlers.ClientCertAuthChallengeHandler;
+import com.microsoft.identity.common.internal.ui.embeddedwebview.challengehandlers.IChallengeCompletionCallback;
+import com.microsoft.identity.common.internal.ui.embeddedwebview.challengehandlers.PKeyAuthChallengeHandler;
+import com.microsoft.identity.common.internal.util.StringUtil;
+
+import java.util.HashMap;
+import java.util.Locale;
+import java.util.Map;
+
+/**
+ * For web view client, we do not distinguish V1 from V2.
+ * Thus we name V1 and V2 webview client as AADWebViewClient, synced with the naming in the iOS common library.
+ * <p>
+ * The only differences between V1 and V2 is
+ * 1. on the start url construction, which is handled in the Authorization request classes.
+ * 2. the auth result is handled in the Authorization result classes.
+ */
+public class AzureActiveDirectoryWebViewClient extends OAuth2WebViewClient {
+    private static final String TAG = AzureActiveDirectoryWebViewClient.class.getSimpleName();
+    private static final String INSTALL_URL_KEY = "app_link";
+
+    AzureActiveDirectoryWebViewClient(@NonNull final Activity activity,
+                                      @NonNull final MicrosoftAuthorizationRequest request,
+                                      @NonNull final IChallengeCompletionCallback callback) {
+        super(activity, request, callback);
+    }
+
+    /**
+     * Give the host application a chance to take over the control when a new url is about to be loaded in the current WebView.
+     * This method was deprecated in API level 24.
+     *
+     * @param view The WebView that is initiating the callback.
+     * @param url  The url to be loaded.
+     * @return return true means the host application handles the url, while return false means the current WebView handles the url.
+     */
+    @Override
+    public boolean shouldOverrideUrlLoading(final WebView view, final String url) {
+        if (StringUtil.isEmpty(url)) {
+            throw new IllegalArgumentException("Redirect to empty url in web view.");
+        }
+        return handleUrl(view, url);
+    }
+
+    /**
+     * Give the host application a chance to take over the control when a new url is about to be loaded in the current WebView.
+     * This method is added in API level 24.
+     *
+     * @param view    The WebView that is initiating the callback.
+     * @param request Object containing the details of the request.
+     * @return return true means the host application handles the url, while return false means the current WebView handles the url.
+     */
+    @Override
+    @RequiresApi(Build.VERSION_CODES.N)
+    public boolean shouldOverrideUrlLoading(final WebView view, final WebResourceRequest request) {
+        final Uri requestUrl = request.getUrl();
+        return handleUrl(view, requestUrl.toString());
+    }
+
+    private boolean handleUrl(final WebView view, final String url) {
```

Comment:
Aside from the critique about the challenge handlers, overall this method is pretty clean and readable 👍